### ISSUE TITLE
Custom Elements TypeScript typings

### ISSUE THREAD

**Issue by marvinhagemeister** [ionic-team/stencil#121]
One great thing about other frameworks based around jsx like react or preact is the validation of the props for custom components via typescript.

Currently in stencil TS is not aware of the shapes of custom components and they are declared with a simple record `Record<string, any>`.

Tried a bit fiddling around with extending the `JSX` namespace but haven't had any luck so far

**Comment by adamdbradley** [comment]
Absolutely agree. I think what we can do is have each component add to the JSX interface: https://github.com/ionic-team/stencil/blob/5b8c54a413960610a831d4ff4a70333a47d9ff21/src/util/jsx-interfaces.ts#L10

But I would love to chat over all ideas on the best way to do this. What's great is that since we do have a build-time compiler, and we can generate d.ts files on-demand, and we are using typescript at the core, we'll hopefully be able to come up with a great solution.

**Comment by marvinhagemeister** [comment]
Yeah, adding component interfaces by hand works, so this should be doable 💯

**Comment by adamdbradley** [comment]
For reference:

```declare global {

  interface ElementTagNameMap {
    "ion-fab": {
      show: boolean;
      method(): void;
    };
  }

  namespace JSX {

    export interface IntrinsicElements {
      ['ion-fab']: {...};
    }

  }
}
```

**Comment by jthoms1** [comment]
I have added Typings for our Custom Elements.  They should now appear in JSX and if you select based on the tagname in vanilla JS.

This is resolved with commit 5f7371b8332163d5ab818ffe0241f5e3d5c5058e.
